File: demo-apps/demoapp002-python-pytest/tooling/performance/benchmark.py
from __future__ import annotations

import csv
import json
import math
import os
import statistics
import time
from pathlib import Path
from typing import Callable

from app_src import Puzzle, PuzzleLoader, SudokuOrchestrator, SudokuSolver

STACK = "DEMOAPP002"
ITERATIONS = int(os.environ.get("BENCHMARK_ITERATIONS", "10"))
WARMUP_ITERATIONS = int(os.environ.get("BENCHMARK_WARMUP_ITERATIONS", "2"))
PUZZLES = ["Easy Scan Grid", "Logic Squeeze Grid", "Empty Grid"]


def main() -> None:
    records: list[dict] = []

    records.append(measure("puzzle-loader", "all", load_puzzles))

    loader = PuzzleLoader("../puzzles.json")
    for puzzle_name in PUZZLES:
        puzzle = loader.get_puzzle_by_name(puzzle_name)
        if puzzle is None:
            raise RuntimeError(f"Puzzle not found: {puzzle_name}")
        records.append(measure("orchestrator-solve", puzzle_name, lambda p=puzzle: solve_puzzle(p)))

    write_results(records)


def load_puzzles() -> None:
    loader = PuzzleLoader("../puzzles.json")
    if loader.get_puzzle_count() == 0:
        raise RuntimeError("No puzzles loaded")


def solve_puzzle(puzzle: Puzzle) -> None:
    solver = SudokuSolver(puzzle.name, puzzle.grid)
    orchestrator = SudokuOrchestrator(solver)
    orchestrator.solve()


def measure(benchmark_name: str, puzzle_name: str, action: Callable[[], None]) -> dict:
    for _ in range(WARMUP_ITERATIONS):
        action()

    samples: list[float] = []
    started = time.perf_counter()
    for _ in range(ITERATIONS):
        sample_started = time.perf_counter()
        action()
        samples.append((time.perf_counter() - sample_started) * 1000)
    duration_ms = (time.perf_counter() - started) * 1000

    mean_ms = statistics.fmean(samples)
    return {
        "stack": STACK,
        "benchmarkName": benchmark_name,
        "puzzleName": puzzle_name,
        "iterations": ITERATIONS,
        "warmupIterations": WARMUP_ITERATIONS,
        "durationMs": duration_ms,
        "meanMs": mean_ms,
        "medianMs": statistics.median(samples),
        "minMs": min(samples),
        "maxMs": max(samples),
        "stddevMs": math.sqrt(statistics.fmean((sample - mean_ms) ** 2 for sample in samples)),
    }


def write_results(records: list[dict]) -> None:
    timestamp = time.strftime("%Y%m%dT%H%M%SZ", time.gmtime())
    out_dir = Path.cwd().parents[1] / ".results" / "performance" / STACK
    out_dir.mkdir(parents=True, exist_ok=True)

    json_path = out_dir / f"performance-{timestamp}.json"
    csv_path = out_dir / f"performance-{timestamp}.csv"
    md_path = out_dir / f"performance-{timestamp}.md"

    json_path.write_text(
        json.dumps({"generatedAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()), "records": records}, indent=2),
        encoding="utf-8",
    )

    with csv_path.open("w", encoding="utf-8", newline="") as handle:
        writer = csv.DictWriter(handle, fieldnames=list(records[0].keys()))
        writer.writeheader()
        writer.writerows(records)

    md_path.write_text(to_markdown(records), encoding="utf-8")
    print(f"Performance results written: {json_path}")


def to_markdown(records: list[dict]) -> str:
    lines = [
        "# DEMOAPP002 Performance Results",
        "",
        "| Benchmark | Puzzle | Iterations | Mean ms | Median ms | Min ms | Max ms | Stddev ms |",
        "|-----------|--------|------------|---------|-----------|--------|--------|-----------|",
    ]
    for record in records:
        lines.append(
            f"| {record['benchmarkName']} | {record['puzzleName']} | {record['iterations']} | "
            f"{record['meanMs']:.3f} | {record['medianMs']:.3f} | {record['minMs']:.3f} | "
            f"{record['maxMs']:.3f} | {record['stddevMs']:.3f} |"
        )
    lines.extend(["", "Reporting mode only: no threshold gate is applied."])
    return "\n".join(lines)


if __name__ == "__main__":
    main()

File: demo-apps/demoapp001-typescript-cypress/tooling/performance/benchmark.ts
import * as fs from 'fs';
import * as path from 'path';
import request from 'supertest';
import { performance } from 'perf_hooks';
import { Puzzle, PuzzleLoader } from '../../app_src/PuzzleLoader';
import { SudokuOrchestrator } from '../../app_src/SudokuOrchestrator';
import { SudokuSolver } from '../../app_src/SudokuSolver';
import { createApp } from '../../app_src/server/app';

interface BenchmarkRecord {
  stack: string;
  benchmarkName: string;
  puzzleName: string;
  iterations: number;
  warmupIterations: number;
  durationMs: number;
  meanMs: number;
  medianMs: number;
  minMs: number;
  maxMs: number;
  stddevMs: number;
}

const STACK = 'DEMOAPP001';
const ITERATIONS = Number(process.env.BENCHMARK_ITERATIONS ?? '10');
const WARMUP_ITERATIONS = Number(process.env.BENCHMARK_WARMUP_ITERATIONS ?? '2');
const PUZZLES = ['Easy Scan Grid', 'Logic Squeeze Grid', 'Empty Grid'];

async function main(): Promise<void> {
  const records: BenchmarkRecord[] = [];

  records.push(
    await measure('puzzle-loader', 'all', () => {
      const loader = new PuzzleLoader('../puzzles.json');
      if (loader.getPuzzleCount() === 0) throw new Error('No puzzles loaded');
    })
  );

  const loader = new PuzzleLoader('../puzzles.json');
  for (const puzzleName of PUZZLES) {
    const puzzle = loader.getPuzzleByName(puzzleName);
    if (!puzzle) throw new Error(`Puzzle not found: ${puzzleName}`);
    records.push(await measure('orchestrator-solve', puzzleName, () => solvePuzzle(puzzle)));
  }

  const app = createApp();
  records.push(
    await measure('api-visualise', 'Easy Scan Grid', async () => {
      await request(app).get('/api/visualise/Easy%20Scan%20Grid').expect(200);
    })
  );

  writeResults(records);
}

async function measure(
  benchmarkName: string,
  puzzleName: string,
  action: () => void | Promise<void>
): Promise<BenchmarkRecord> {
  for (let i = 0; i < WARMUP_ITERATIONS; i++) {
    await action();
  }

  const samples: number[] = [];
  const started = performance.now();
  for (let i = 0; i < ITERATIONS; i++) {
    const sampleStarted = performance.now();
    await action();
    samples.push(performance.now() - sampleStarted);
  }
  const durationMs = performance.now() - started;

  return toRecord(benchmarkName, puzzleName, durationMs, samples);
}

function solvePuzzle(puzzle: Puzzle): void {
  const solver = new SudokuSolver(puzzle.name, puzzle.grid);
  const orchestrator = new SudokuOrchestrator(solver);
  orchestrator.solve();
}

function toRecord(
  benchmarkName: string,
  puzzleName: string,
  durationMs: number,
  samples: number[]
): BenchmarkRecord {
  const sorted = [...samples].sort((a, b) => a - b);
  const mean = samples.reduce((sum, value) => sum + value, 0) / samples.length;
  const variance = samples.reduce((sum, value) => sum + (value - mean) ** 2, 0) / samples.length;

  return {
    stack: STACK,
    benchmarkName,
    puzzleName,
    iterations: ITERATIONS,
    warmupIterations: WARMUP_ITERATIONS,
    durationMs,
    meanMs: mean,
    medianMs: sorted[Math.floor(sorted.length / 2)],
    minMs: sorted[0],
    maxMs: sorted[sorted.length - 1],
    stddevMs: Math.sqrt(variance),
  };
}

function writeResults(records: BenchmarkRecord[]): void {
  const timestamp = new Date()
    .toISOString()
    .replace(/[-:]/g, '')
    .replace(/\.\d{3}Z$/, 'Z');
  const outDir = path.resolve(process.cwd(), '../../.results/performance', STACK);
  fs.mkdirSync(outDir, { recursive: true });

  const jsonPath = path.join(outDir, `performance-${timestamp}.json`);
  const csvPath = path.join(outDir, `performance-${timestamp}.csv`);
  const mdPath = path.join(outDir, `performance-${timestamp}.md`);

  fs.writeFileSync(
    jsonPath,
    JSON.stringify({ generatedAt: new Date().toISOString(), records }, null, 2)
  );
  fs.writeFileSync(csvPath, toCsv(records));
  fs.writeFileSync(mdPath, toMarkdown(records));

  console.log(`Performance results written: ${jsonPath}`);
}

function toCsv(records: BenchmarkRecord[]): string {
  const columns = [
    'stack',
    'benchmarkName',
    'puzzleName',
    'iterations',
    'warmupIterations',
    'durationMs',
    'meanMs',
    'medianMs',
    'minMs',
    'maxMs',
    'stddevMs',
  ];
  return [
    columns.join(','),
    ...records.map((record) =>
      columns
        .map((column) => JSON.stringify(record[column as keyof BenchmarkRecord] ?? ''))
        .join(',')
    ),
  ].join('\n');
}

function toMarkdown(records: BenchmarkRecord[]): string {
  return [
    '# DEMOAPP001 Performance Results',
    '',
    '| Benchmark | Puzzle | Iterations | Mean ms | Median ms | Min ms | Max ms | Stddev ms |',
    '|-----------|--------|------------|---------|-----------|--------|--------|-----------|',
    ...records.map(
      (record) =>
        `| ${record.benchmarkName} | ${record.puzzleName} | ${record.iterations} | ${format(record.meanMs)} | ${format(record.medianMs)} | ${format(record.minMs)} | ${format(record.maxMs)} | ${format(record.stddevMs)} |`
    ),
    '',
    'Reporting mode only: no threshold gate is applied.',
  ].join('\n');
}

function format(value: number): string {
  return value.toFixed(3);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
